### ISSUE TITLE
[TW#19673] aws_iot: failed! mbedtls_ssl_handshake returned -0x4310 (IDFGH-766)

### ISSUE THREAD

**Issue by satishIoT** [espressif/esp-idf#1750]
Hi 
I keep facing this error now and then when ever i work with shadow.
1) aws_iot: failed! mbedtls_ssl_handshake returned -0x4310
    E (18143) APP_DEBUG: aws_iot_shadow_connect returned error -4, aborting...
2) Shadow update timeout .. 
3)  And if it succeeds posting messages then again at some point i get 
aws_iot: failed! mbedtls_ssl_handshake returned -0x4310
and never able to reconnect or recover from there on.
4) This certificate has no flags : what this means will this cause any issue w.r.t connection to aws. 
pl advice !!
sats

Complete log file .....................................................

Rebooting...
ets Jun  8 2016 00:22:57

rst:0xc (SW_CPU_RESET),boot:0x17 (SPI_FAST_FLASH_BOOT)
configsip: 0, SPIWP:0xee
clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
mode:DIO, clock div:2
load:0x3fff0018,len:4
load:0x3fff001c,len:5708
load:0x40078000,len:0
load:0x40078000,len:13804
entry 0x40079030
I (68) boot: Detected ESP32
I (32) boot: ESP-IDF v3.0-dev-1392-g579201bc-dirty 2nd stage bootloader
I (32) boot: compile time 17:09:35
I (33) boot: Enabling RNG early entropy source...
I (39) boot: SPI Speed      : 40MHz
I (43) boot: SPI Mode       : DIO
I (47) boot: SPI Flash Size : 4MB
I (51) boot: Partition Table:
I (55) boot: ## Label            Usage          Type ST Offset   Length
I (62) boot:  0 nvs              WiFi data        01 02 00009000 00006000
I (69) boot:  1 phy_init         RF data          01 01 0000f000 00001000
I (77) boot:  2 factory          factory app      00 00 00010000 00200000
I (84) boot: End of partition table
I (88) esp_image: segment 0: paddr=0x00010020 vaddr=0x3f400020 size=0x4a5b0 (304                                                                                        560) map
I (203) esp_image: segment 1: paddr=0x0005a5d8 vaddr=0x3ffc0000 size=0x047f4 ( 1                                                                                        8420) load
I (211) esp_image: segment 2: paddr=0x0005edd4 vaddr=0x40080000 size=0x00400 (                                                                                          1024) load
I (211) esp_image: segment 3: paddr=0x0005f1dc vaddr=0x40080400 size=0x00e34 (                                                                                          3636) load
I (220) esp_image: segment 4: paddr=0x00060018 vaddr=0x400d0018 size=0xd794c (88                                                                                        3020) map
I (534) esp_image: segment 5: paddr=0x0013796c vaddr=0x40081234 size=0x14d9c ( 8                                                                                        5404) load
I (570) esp_image: segment 6: paddr=0x0014c710 vaddr=0x400c0000 size=0x00064 (                                                                                           100) load
I (570) esp_image: segment 7: paddr=0x0014c77c vaddr=0x50000000 size=0x00008 (                                                                                             8) load
I (589) boot: Loaded app from partition at offset 0x10000
I (589) boot: Disabling RNG early entropy source...
I (590) cpu_start: Pro cpu up.
I (594) cpu_start: Starting app cpu, entry point is 0x400810f8
I (584) cpu_start: App cpu up.
I (605) heap_init: Initializing. RAM available for dynamic allocation:
D (611) heap_init: New heap initialised at 0x3ffaff10
I (616) heap_init: At 3FFAFF10 len 000000F0 (0 KiB): DRAM
D (622) heap_init: New heap initialised at 0x3ffd4fa8
I (628) heap_init: At 3FFD4FA8 len 0000B058 (44 KiB): DRAM
I (634) heap_init: At 3FFE0440 len 00003BC0 (14 KiB): D/IRAM
I (640) heap_init: At 3FFE4350 len 0001BCB0 (111 KiB): D/IRAM
D (647) heap_init: New heap initialised at 0x40095fd0
I (652) heap_init: At 40095FD0 len 0000A030 (40 KiB): IRAM
I (658) cpu_start: Pro cpu start user code
D (670) clk: RTC_SLOW_CLK calibration value: 3293798
D (9) intr_alloc: Connected src 57 to int 2 (cpu 0)
D (9) intr_alloc: Connected src 24 to int 3 (cpu 0)
I (10) cpu_start: Starting scheduler on PRO CPU.
D (0) intr_alloc: Connected src 25 to int 2 (cpu 1)
I (0) cpu_start: Starting scheduler on APP CPU.
D (24) heap_init: New heap initialised at 0x3ffe0440
D (34) heap_init: New heap initialised at 0x3ffe4350
D (44) intr_alloc: Connected src 16 to int 9 (cpu 0)
I (44) APP_DEBUG: Hello world!
I (44) APP_DEBUG: This is ESP32 chip with 2 CPU cores, WiFi/BT/BLE,
I (54) APP_DEBUG: silicon revision 1,
I (64) APP_DEBUG: 4MB external flash

I (64) APP_DEBUG: Version: ECO_V0.04-INT

D (74) intr_alloc: Connected src 49 to int 12 (cpu 0)
D (2074) intr_alloc: Connected src 14 to int 13 (cpu 0)
I (2074) APP_DEBUG: Timer ISR init is done
D (2074) intr_alloc: Connected src 34 to int 17 (cpu 0)
I (2074) APP_DEBUG: Not in MF mode
D (2084) nvs: nvs_flash_init_custom partition=nvs start=9 count=6
D (2124) nvs: nvs_open_from_partition storage 1
D (2124) nvs: nvs_get_str_or_blob block1
D (2124) nvs: nvs_get_str_or_blob block1
I (2124) APP_DEBUG: retriving from EEPROM size = 328
D (2124) nvs: nvs_get_str_or_blob block1_crc
D (2134) nvs: nvs_get_str_or_blob block1_crc
I (2134) APP_DEBUG: EepromInit(698) checkSum = 0x17e949bc
I (2144) APP_DEBUG: Checksum Passed
I (2144) APP_DEBUG: Not a deep sleep reset
I (2154) APP_DEBUG: Running Model is: EWS ECR3700R20
I (2154) APP_DEBUG: Shadow not connected prop_tbl_offset=85

I (2164) APP_DEBUG: Shadow not connected prop_tbl_offset=86

I (2174) gpio: GPIO[27]| InputEn: 0| OutputEn: 1| OpenDrain: 0| Pullup: 0| Pulld                                                                                        own: 0| Intr:0
D (2184) intr_alloc: Connected src 30 to int 18 (cpu 0)
I (2184) APP_DEBUG: Shadow not connected prop_tbl_offset=84

I (2194) APP_DEBUG: Flag.Valve_Pos = 1
I (2194) gpio: GPIO[21]| InputEn: 1| OutputEn: 0| OpenDrain: 0| Pullup: 0| Pulld                                                                                        own: 0| Intr:0
I (2204) gpio: GPIO[33]| InputEn: 1| OutputEn: 0| OpenDrain: 0| Pullup: 0| Pulld                                                                                        own: 0| Intr:0
D (2214) intr_alloc: Connected src 22 to int 19 (cpu 0)
D (2224) BTDM_INIT: .data initialise [0x3ffae6e0] <== [0x4000d890]

D (2224) BTDM_INIT: .bss initialise [0x3ffb0000] - [0x3ffb09a8]

D (2234) BTDM_INIT: .bss initialise [0x3ffb09a8] - [0x3ffb1ddc]

D (2234) BTDM_INIT: .bss initialise [0x3ffb1ddc] - [0x3ffb2730]

D (2244) BTDM_INIT: .bss initialise [0x3ffb2730] - [0x3ffb8000]

D (2254) BTDM_INIT: .bss initialise [0x3ffb8000] - [0x3ffbbb28]

D (2254) BTDM_INIT: .bss initialise [0x3ffbbb28] - [0x3ffbdb28]

D (2264) BTDM_INIT: .bss initialise [0x3ffbdb28] - [0x3ffc0000]

I (2264) system_api: Base MAC address is not set, read default base MAC address                                                                                         from BLK0 of EFUSE
D (2514) phy_init: loading PHY init data from application binary
D (2514) nvs: nvs_open_from_partition phy 0
D (2514) nvs: nvs_get cal_version 4
D (2514) nvs: nvs_get_str_or_blob cal_mac
D (2524) nvs: nvs_get_str_or_blob cal_data
D (2554) nvs: nvs_close 2
I (2584) phy: phy_version: 366.0, ba9923d, Oct 31 2017, 18:06:17, 0, 0
I (2584) APP_DEBUG: ble_task init bluetooth
D (2584) nvs: nvs_open_from_partition bt_config.conf 1
D (2584) nvs: nvs_get_str_or_blob bt_cfg_key
D (2594) nvs: nvs_close 3
D (2594) nvs: nvs_open_from_partition bt_config.conf 1
D (2594) nvs: nvs_set_blob bt_cfg_key 216
D (2604) nvs: nvs_close 4
I (2634) APP_DEBUG: create attribute table successfully, the number handle = 9
I (2644) APP_DEBUG: SERVICE_START_EVT, status 0, service_handle 40

I (2644) APP_DEBUG: advertising start successfully
I (3174) APP_DEBUG: Time in seconds, (started from bootup) = 0
I (3174) APP_DEBUG: Current Time 00:18:07
I (4174) APP_DEBUG: Time in seconds, (started from bootup) = 1
I (4174) APP_DEBUG: Current Time 00:18:08
I (4184) APP_DEBUG: esphal_getAdc1Voltage(266) *voltage = 421
I (4184) APP_DEBUG: Shadow not connected prop_tbl_offset=90

I (4194) APP_DEBUG: Shadow not connected prop_tbl_offset=89

I (4194) APP_DEBUG: Cust.Salt_Level = 10
I (4204) APP_DEBUG: Get_out_of_salt_days = 34
D (4224) nvs: nvs_open_from_partition misc 1
D (4224) nvs: nvs_get_str_or_blob log
I (4224) wifi: wifi firmware version: f204566
I (4224) wifi: config NVS flash: enabled
I (4224) wifi: config nano formating: disabled
I (4234) system_api: Base MAC address is not set, read default base MAC address                                                                                         from BLK0 of EFUSE
I (4244) system_api: Base MAC address is not set, read default base MAC address                                                                                         from BLK0 of EFUSE
D (4254) nvs: nvs_open_from_partition nvs.net80211 1
D (4254) nvs: nvs_get opmode 1
D (4254) nvs: nvs_get_str_or_blob sta.ssid
D (4264) nvs: nvs_get_str_or_blob sta.mac
D (4264) nvs: nvs_get sta.authmode 1
D (4274) nvs: nvs_get_str_or_blob sta.pswd
D (4274) nvs: nvs_get_str_or_blob sta.pmk
D (4274) nvs: nvs_get sta.chan 1
D (4284) nvs: nvs_get auto.conn 1
D (4284) nvs: nvs_get bssid.set 1
D (4284) nvs: nvs_get_str_or_blob sta.bssid
D (4294) nvs: nvs_get sta.phym 1
D (4294) nvs: nvs_get sta.phybw 1
D (4294) nvs: nvs_get_str_or_blob sta.apsw
D (4304) nvs: nvs_get_str_or_blob sta.apinfo
D (4304) nvs: nvs_get sta.scan_method 1
D (4304) nvs: nvs_get sta.sort_method 1
D (4304) nvs: nvs_get sta.minrssi 1
D (4304) nvs: nvs_get sta.minauth 1
D (4314) nvs: nvs_get_str_or_blob ap.ssid
D (4314) nvs: nvs_get_str_or_blob ap.mac
D (4324) nvs: nvs_get_str_or_blob ap.passwd
D (4324) nvs: nvs_get_str_or_blob ap.pmk
D (4324) nvs: nvs_get ap.chan 1
D (4334) nvs: nvs_get ap.authmode 1
D (4334) nvs: nvs_get ap.hidden 1
D (4334) nvs: nvs_get ap.max.conn 1
D (4344) nvs: nvs_get bcn.interval 2
D (4344) nvs: nvs_get ap.phym 1
D (4344) nvs: nvs_get ap.phybw 1
D (4354) nvs: nvs_get ap.sndchan 1
D (4354) nvs: nvs_set_blob sta.mac 6
D (4364) nvs: nvs_set_blob ap.mac 6
I (4364) wifi: Init dynamic tx buffer num: 32
I (4364) wifi: Init data frame dynamic rx buffer num: 32
I (4374) wifi: Init management frame dynamic rx buffer num: 32
I (4374) wifi: wifi driver task: 3ffec828, prio:23, stack:4096
I (4384) wifi: Init static rx buffer num: 10
I (4384) wifi: Init dynamic rx buffer num: 32
I (4394) wifi: wifi power manager task: 0x3fff1468 prio: 21 stack: 2560
I (4394) APP_DEBUG: Setting WiFi configuration SSID FILxxxxxx...
D (4404) RTC_MODULE: Wi-Fi takes adc2 lock.
D (4404) phy_init: loading PHY init data from application binary
D (4414) nvs: nvs_open_from_partition phy 0
D (4414) nvs: nvs_get cal_version 4
D (4424) nvs: nvs_get_str_or_blob cal_mac
D (4424) nvs: nvs_get_str_or_blob cal_data
D (4434) nvs: nvs_close 7
I (4444) wifi: mode : sta (30:ae:a4:38:57:54)
I (4444) APP_DEBUG: AWS IoT SDK Version 2.1.1-
D (4444) event: SYSTEM_EVENT_STA_START
D (4444) tcpip_adapter: check: local, if=0 fn=0x401284a0

D (4454) tcpip_adapter: call api in lwip: ret=0x0, give sem
D (4454) tcpip_adapter: check: remote, if=0 fn=0x401284a0

I (4584) wifi: n:1 0, o:1 0, ap:255 255, sta:1 0, prof:1
I (5574) wifi: state: init -> auth (b0)
I (5574) APP_DEBUG: Time in seconds, (started from bootup) = 2
I (5574) APP_DEBUG: Current Time 00:18:08
I (5574) wifi: state: auth -> assoc (0)
I (5584) wifi: state: assoc -> run (10)
I (5604) wifi: connected with FILxxxxx, channel 1
D (5604) event: SYSTEM_EVENT_STA_CONNECTED, ssid:FILxxxxxx, ssid_len:12, bssi                                                                                        d:74:26:ac:e3:e4:d0, channel:1, authmode:3
D (5614) tcpip_adapter: check: local, if=0 fn=0x401286f0

D (5614) tcpip_adapter: call api in lwip: ret=0x0, give sem
D (5624) tcpip_adapter: check: remote, if=0 fn=0x401286f0

D (5634) tcpip_adapter: check: local, if=0 fn=0x40128ba8

D (5634) tcpip_adapter: dhcp client init ip/mask/gw to all-0
D (5644) tcpip_adapter: if0 start ip lost tmr: enter
D (5644) tcpip_adapter: if0 start ip lost tmr: no need start because netif=0x3ff                                                                                        f1b20 interval=120 ip=0
D (5654) tcpip_adapter: dhcp client start successfully
D (5664) tcpip_adapter: call api in lwip: ret=0x0, give sem
D (5664) tcpip_adapter: check: remote, if=0 fn=0x40128ba8

I (6164) APP_DEBUG: Time in seconds, (started from bootup) = 3
I (6164) APP_DEBUG: Current Time 00:18:10
D (6744) tcpip_adapter: if0 dhcpc cb
D (6744) tcpip_adapter: if0 ip changed=1
D (6744) event: SYSTEM_EVENT_STA_GOT_IP, ip:192.168.2.169, mask:255.255.255.0, g                                                                                        w:192.168.2.1
I (6744) event: sta ip: 192.168.2.169, mask: 255.255.255.0, gw: 192.168.2.1
I (6754) APP_DEBUG: Shadow Init
I (6754) APP_DEBUG: Shadow Connect
D (6764) aws_iot: Seeding the random number generator...
D (6764) aws_iot: Loading embedded CA root certificate ...
D (6774) aws_iot: ok (0 skipped)
D (6774) aws_iot: Loading embedded client certificate...
D (6784) aws_iot: Loading embedded client private key...
D (6954) aws_iot: ok
D (6954) aws_iot: Connecting to a2ng0twleztum6.iot.us-east-1.amazonaws.com/8883.                                                                                        ..
I (7164) APP_DEBUG: Time in seconds, (started from bootup) = 4
I (7164) APP_DEBUG: Current Time 00:18:11
D (7434) aws_iot: ok
D (7434) aws_iot: Setting up the SSL/TLS structure...
D (7434) aws_iot: SSL state connect : 0
D (7434) aws_iot: ok
D (7434) aws_iot: SSL state connect : 0
D (7434) aws_iot: Performing the SSL/TLS handshake...
I (8164) APP_DEBUG: Time in seconds, (started from bootup) = 5
I (8164) APP_DEBUG: Current Time 00:18:12
I (8584) wifi: pm start, type:0

D (8594) aws_iot: Verify requested for (Depth 2):
D (8594) aws_iot: cert. version     : 3
serial number     : 18:DA:D1:9E:26:7D:E8:BB:4A:21:58:CD:CC:6B:3B:4A
issuer name       : C=US, O=VeriSign, Inc., OU=VeriSign Trust Network, OU=(c) 20                                                                                        06 VeriSign, Inc. - For authorized use only, CN=VeriSign Class 3 Public Primary                                                                                         Cert
D (8614) aws_iot:   This certificate has no flags
D (8614) aws_iot: Verify requested for (Depth 1):
D (8624) aws_iot: cert. version     : 3
serial number     : 3F:92:87:BE:9D:1D:A4:A3:7A:9D:F6:28:2E:77:5A:C4
issuer name       : C=US, O=VeriSign, Inc., OU=VeriSign Trust Network, OU=(c) 20                                                                                        06 VeriSign, Inc. - For authorized use only, CN=VeriSign Class 3 Public Primary                                                                                         Cert
D (8644) aws_iot:   This certificate has no flags
D (8654) aws_iot: Verify requested for (Depth 0):
D (8654) aws_iot: cert. version     : 3
serial number     : 0A:84:9B:FA:E6:BB:18:70:3F:F7:76:20:47:7C:7D:90
issuer name       : C=US, O=Symantec Corporation, OU=Symantec Trust Network, CN=                                                                                        Symantec Class 3 ECC 256 bit SSL CA - G2
subject name      : C=US, ST=Washington, L=
D (8674) aws_iot:   This certificate has no flags
I (9164) APP_DEBUG: Time in seconds, (started from bootup) = 6
I (9164) APP_DEBUG: Current Time 00:18:13
I (10164) APP_DEBUG: Time in seconds, (started from bootup) = 7
I (10174) APP_DEBUG: Current Time 00:18:14
I (11164) APP_DEBUG: Time in seconds, (started from bootup) = 8
I (11164) APP_DEBUG: Current Time 00:18:15
I (12164) APP_DEBUG: Time in seconds, (started from bootup) = 9
I (12164) APP_DEBUG: Current Time 00:18:16
Task watchdog got triggered. The following tasks did not reset the watchdog in t                                                                                        ime:
 - IDLE (CPU 0)
Tasks currently running:
CPU 0: main_renesas
CPU 1: aws_iot_task
E (12284) aws_iot: failed! mbedtls_ssl_handshake returned -0x4310
E (12284) APP_DEBUG: aws_iot_shadow_connect returned error -4, aborting...
abort() was called at PC 0x4010cedf on core 1

Backtrace: 0x4008c5cc:0x3ffdfbb0 0x4008c6cb:0x3ffdfbd0 0x4010cedf:0x3ffdfbf0

**Comment by ameyinamdar** [comment]
Satish, the error -0x4310 refers to MBEDTLS_ERR_RSA_PRIVATE_FAILED (0x4300) + MBEDTLS_ERR_MPI_ALLOC_FAILED (0x0010). It looks like your system is running out of memory.

**Comment by satishIoT** [comment]
Any more input im using ERS32 WROOM module and i guess we have enough memory to deal with all tasks. Any changes in menuconfig to be taken care in mbedtls ?

**Comment by mahavirj** [comment]
@satishIoT 
Did you check how much dynamic memory is available (please note that WROOM module does not have external SPIRAM) before starting TLS connection (or at various stages in AWS IoT shadow task)? You can use `esp_get_free_heap_size` and `esp_get_minimum_free_heap_size` APIs for this.

**Comment by negativekelvin** [comment]
https://github.com/espressif/esp-idf/issues/1695

**Comment by satishIoT** [comment]
Thanks @negativekelvin your suggestion was spot on. 
But i have another issue ; im using the shadow example and as understood if n/w disconnects then the autoreconnect of shadow should have happened and should have looped in 
while(NETWORK_ATTEMPTING_RECONNECT == rc || NETWORK_RECONNECTED == rc || SUCCESS == rc) {
        rc = aws_iot_shadow_yield(&mqttClient, 200);
        if(NETWORK_ATTEMPTING_RECONNECT == rc || shadowUpdateInProgress) {
            rc = aws_iot_shadow_yield(&mqttClient, 1000);
            // If the client is attempting to reconnect, or already waiting on a shadow update,
            // we will skip the rest of the loop.
            continue;
        }
But instead it goes all down through the loop and give rc = -28. 
Pl suggest how to initiate the reconnect when ever n/w breakdown happens and comes back. 

thanks 
satish

**Comment by satishIoT** [comment]
![image](https://user-images.githubusercontent.com/26374917/37966078-ea30dd7a-31e4-11e8-901b-8fa4b426f634.png)

pl suggest on above errors.

**Comment by satishIoT** [comment]
Dear Experts , Awaiting for  suggestions on how to deal with above errors , to make my connection robust .

thanks in advance !!

**Comment by ameyinamdar** [comment]
@satishIoT "mbedtls_net_connect returned -0x52" indicates hostname resolution failure (internet connectivity issue?)

**Comment by satishIoT** [comment]
@ameyinamdar - Im using hotspot to verify/test how the connectivity works say if i try posting mqtt messages say at every minute then i dont see such issue but if i dont enable my hotspot for 45 min or more ( kind of connection break )  i see this error along with -14 and -13 error also. i want to have a robust method to connect my device when ever connectivity returns back. what should i need to take care in such cases.

**Comment by satishIoT** [comment]
Dear Experts , Awaiting for suggestions on how to deal with above errors , to make my connection robust .

thanks in advance !!

**Comment by satishIoT** [comment]
Dear Experts , Awaiting for suggestions on how to deal with above errors , to make my connection robust .

thanks in advance !!

**Comment by Alvin1Zhang** [comment]
@satishIoT Hi, sorry that get back to you so late. Does your issue still exist? Thanks.

**Comment by mahavirj** [comment]
@satishIoT Closing this issue (not reproducible at our end), in case it still exists please feel free to reopen at https://github.com/espressif/esp-aws-iot (we moved `aws-iot` to independent repository)
